scripts/build/builders/esp32.py

```diff
@@ -0,0 +1,93 @@
+import logging
+import os
+from enum import Enum, auto
+
+from .builder import Builder
+
+
+class Esp32Board(Enum):
+  DevKitC = auto()
+  M5Stack = auto()
+
+
+class Esp32App(Enum):
+  ALL_CLUSTERS = auto()
+  LOCK = auto()
+
+  @property
+  def ExampleName(self):
+    if self == Esp32App.ALL_CLUSTERS:
+      return 'all-clusters-app'
+    elif self == Esp32App.LOCK:
+      return 'lock-app'
+    else:
+      raise Exception('Unknown app type: %r' % self)
+
+  @property
+  def AppNamePrefix(self):
+    if self == Esp32App.ALL_CLUSTERS:
+      return 'chip-all-clusters-app'
+    elif self == Esp32App.LOCK:
+      return 'chip-lock-app'
+    else:
+      raise Exception('Unknown app type: %r' % self)
+
+
+def DefaultsFileName(board: Esp32Board, app: Esp32App):
+  if app != Esp32App.ALL_CLUSTERS:
+    # only all-clusters has a specific defaults name
+    return None
+
+  if board == Esp32Board.DevKitC:
+    return 'sdkconfig_devkit.defaults'
+  elif board == Esp32Board.M5Stack:
+    return 'sdkconfig_m5stack.defaults'
+  else:
+    raise Exception('Unknown board type')
+
+
+class Esp32Builder(Builder):
+
+  def __init__(self,
+               root,
+               output_dir: str,
+               board: Esp32Board = Esp32Board.M5Stack,
+               app: Esp32App = Esp32App.ALL_CLUSTERS):
+    super(Esp32Builder, self).__init__(root, output_dir)
+    self.board = board
+    self.app = app
+
+  def _IdfEnvExecute(self, cmd, **kargs):
+    self._Execute(
+        ['bash', '-c', 'source $IDF_PATH/export.sh; %s' % cmd], **kargs)
+
+  def generate(self):
+    if os.path.exists(os.path.join(self.output_dir, 'build.ninja')):
+      return
+
+    defaults = DefaultsFileName(self.board, self.app)
+
+    cmd = 'idf.py'
+
+    if defaults:
+      cmd += " -D SDKCONFIG_DEFAULTS='%s'" % defaults
+
+    cmd += ' -C examples/%s/esp32 -B %s reconfigure' % (self.app.ExampleName,
```

Comment:
Please use shlex here for the requisite escaping.

Alternatively, it's usually best to use the subprocess module directly as a list, but we'd need a run_in_environment.sh for that.

Reply:
Done for te output (I believe the examples we can ensure no escaping needed)